tests/abstract_model_inheritance/tests.py

```diff
@@ -0,0 +1,320 @@
+        self.assertEqual(Cousin._meta.get_field('name').max_length, 50)
+        self.assertEqual(GrandChild._meta.get_field('name').max_length, 50)
+
+    def test_multiple_parents_mro(self):
+        class GrandMother(models.Model):
+            class Meta:
+                abstract = True
+
+        class GrandFather(models.Model):
+            name = models.CharField(max_length=30)
+
+            class Meta:
+                abstract = True
+
+        class Mother(GrandMother, GrandFather):
+            class Meta:
+                abstract = True
+
+        class Father(GrandMother, GrandFather):
+            name = models.CharField(max_length=50)
+
+            class Meta:
+                abstract = True
+
+        class Child(Mother, Father):
+            pass
+
+        self.assertEqual(Mother._meta.get_field('name').max_length, 30)
+        self.assertEqual(Father._meta.get_field('name').max_length, 50)
+        self.assertEqual(Child._meta.get_field('name').max_length, 50)
+
+    def test_multiple_inheritance_cannot_shadow_concrete_inherited_field(self):
+
+        class ConcreteParent(models.Model):
+            name = models.CharField(max_length=255)
+
+        class AbstractParent(models.Model):
+            name = models.IntegerField()
+
+            class Meta:
+                abstract = True
+
+        class Child(ConcreteParent, AbstractParent):
+            pass
+
+        class AnotherChild(AbstractParent, ConcreteParent):
+            pass
+
+        errors = AnotherChild.check()
+        expected = [
+            Error(
+                "The field 'name' clashes with the field 'name' from model 'abstract_model_inheritance.concreteparent'.",
+                hint=None,
+                obj=AnotherChild._meta.get_field('name'),
+                id="models.E006",
+            ),
+        ]
+
+        self.assertEquals(Child._meta.get_field('name').get_internal_type(), 'CharField')
+        self.assertEquals(errors, expected)
+
+    def test_virtual_field(self):
+        class RelationModel(models.Model):
+            content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+            object_id = models.PositiveIntegerField()
+            content_object = GenericForeignKey('content_type', 'object_id')
+
+        class RelatedModelAbstract(models.Model):
+            field = GenericRelation(RelationModel)
+
+            class Meta:
+                abstract = True
+
+        class ModelAbstract(models.Model):
+            field = models.CharField(max_length=100)
+
+            class Meta:
+                abstract = True
+
+        class OverrideRelatedModelAbstract(RelatedModelAbstract):
+            field = models.CharField(max_length=100)
+
+        class ExtendModelAbstract(ModelAbstract):
+            field = GenericRelation(RelationModel)
+
+        self.assertTrue(isinstance(OverrideRelatedModelAbstract._meta.get_field('field'), models.CharField))
```

Comment:
`assertIsInstance` (if there's a reason for using that here and `get_internal_type()` above, it's not obvious to me and could probably use a comment)